254 Fix Comments XSS Vulnerability

LinasRaj opened the pull request:
Fixed persistent XSS vulnerability in comments by implementing HTML sanitization with bleach library.
Added comprehensive security test suite covering XSS attack vectors.

LinasRaj committed:
Fix WEB-1: Persistent XSS in comments with bleach sanitization

LinasRaj committed:
Fix Python 3.11 typing regression in orgs/services.py (from #1888)

karina-klinkeviciute commented on vitrina/orgs/services.py:
```diff
@@ -79,7 +79,9 @@ class Role(Enum):
 ACL_RULE = tuple[type[Model], Action]
-EXISTING_DATASET_ACL_RULE = tuple[Union[DATASET_RELATED_OBJECTS], bool, str, Action]
+EXISTING_DATASET_ACL_RULE = tuple[
+    Union[Dataset, DatasetDistribution, DatasetStructure, DatasetRelation, Request], bool, str, Action
```
o kodėl čia pakeitei? Kaip matau, išimtas `DatasetAttribution`

LinasRaj commented:
Metė klaidą, nepraeidinėjo testų prieš mano pakeitimus.

-------- Original Message --------
On 10/7/25 21:08, Karina  wrote:

> @karina-klinkeviciute commented on this pull request.
>
> ---------------------------------------------------------------
>
> In [vitrina/orgs/services.py](https://github.com/atviriduomenys/katalogas/pull/1989#discussion_r2411487725):
>
>> @@ -79,7 +79,9 @@ class Role(Enum):
>
>  DATASET_IS_PUBLIC = True
>  ACL_RULE = tuple[type[Model], Action]
> -EXISTING_DATASET_ACL_RULE = tuple[Union[DATASET_RELATED_OBJECTS], bool, str, Action]
> +EXISTING_DATASET_ACL_RULE = tuple[
> +    Union[Dataset, DatasetDistribution, DatasetStructure, DatasetRelation, Request], bool, str, Action
>
> o kodėl čia pakeitei? Kaip matau, išimtas Dataset_attribution
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/atviriduomenys/katalogas/pull/1989#pullrequestreview-3311332995), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/BXY5RNJCU4YP7NBMUHW4II33WP6THAVCNFSM6AAAAACIQ6M4Y6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTGMJRGMZTEOJZGU).
> You are receiving this because you authored the thread.Message ID: ***@***.***>

karina-klinkeviciute commented on tests/security/test_comments_xss.py:
These tests are very comprehensive, but it would be also useful to test the final result. You can do it by using Django test client. Example:

```python
@pytest.mark.django_db
def test_home_page(client):
    response = client.get('/')
    assert response.status_code == 200
    assert b"Sveiki" in response.content
```

karina-klinkeviciute reviewed (commented):
<img width="893" height="680" alt="Image" src="https://github.com/user-attachments/assets/f12235d7-0456-4517-a4af-927cb4a834f3" /> komentaro `parent` taip pat gali būti komentaras, tai `markdown` filtras gerai būtų ir čia (nes ir 20 simbolių gali užtekti trumpam scriptui, manau)

https://github.com/atviriduomenys/katalogas/blob/c7b4d883de1299533b9811a0f030efdcdd01a3a8/vitrina/templates/component/comment.html 


https://test.data.gov.lt/projects/434

karina-klinkeviciute commented on vitrina/orgs/services.py:
```diff
@@ -79,7 +79,9 @@ class Role(Enum):
 DATASET_IS_PUBLIC = True
 ACL_RULE = tuple[type[Model], Action]
-EXISTING_DATASET_ACL_RULE = tuple[Union[DATASET_RELATED_OBJECTS], bool, str, Action]
+EXISTING_DATASET_ACL_RULE = tuple[
```
komentarai dar gali būti prie panaudojimo atvejų: https://test.data.gov.lt/projects/435 (jei čia žymima, kur gali būti komentarai)

Struchid commented on vitrina/security.py:
```diff
@@ -0,0 +1,122 @@
+ALLOWED_PROTOCOLS = ["http", "https", "mailto"]
+
+
+def sanitize_html(html_content: str, strip=False) -> str:
```
```suggestion
def sanitize_html(html_content: str, strip: bool = False) -> str:
```

Struchid commented on tests/security/test_comments_xss.py:
```diff
@@ -0,0 +1,203 @@
+]
+
+
+@pytest.fixture
+def user(db):
+    """Create test user."""
+    return User.objects.create_user(
+        email='test@example.com',
+        first_name='Test',
+        last_name='User',
+        password='testpass123'
+    )
+
+
+@pytest.fixture
+def organization(db):
+    """Create test organization."""
+    return Organization.add_root(
+        title='Test Organization',
+        company_code='123456'
+    )
+
+
+@pytest.fixture
+def dataset(db, organization):
+    """Create test dataset."""
+    return Dataset.objects.create(
+        title='Test Dataset',
+        organization=organization
+    )
```
**Optional:** I'd keep fixtures in `conftest.py` so that they would not lay around the tests

Struchid commented on tests/security/test_comments_xss.py:
```diff
@@ -0,0 +1,203 @@
+    
+    # Dangerous content should be escaped (not executable)
+    # Check that we don't have UNESCAPED dangerous tags
+    assert '<script>' not in html_lower, f"Unescaped script tag found in: {html}"
```
Would be nice to keep consistent with the quotation marks, either singles or doubles. E.g.:

```suggestion
    assert "<script>" not in html_lower, f"Unescaped script tag found in: {html}"
```

Obviously, not in places where you need a string that has double quotes inside it.

LinasRaj committed:
Address PR feedback: move fixtures to conftest, add ARIA support, enhance tests add end to end, fix comment parent, additional improvements.

LinasRaj commented:
Thanks for the thorough review. I've addressed the feedback in the latest commit: https://github.com/atviriduomenys/katalogas/pull/1989/commits/cf3bced5c959b925523fd467451ad0a3678ecd62

## Implemented in This PR

1. @Struchid - Moved fixtures (user, organization, dataset) to tests/conftest.py as suggested. Also added temp_media_root fixture to prevent permission errors when running tests locally.

2. @Struchid - Added type hint: strip: bool = False

3. @Struchid - Fixed quote consistency (double quotes by default, single quotes only when string contains double quotes inside)

4. @karina-klinkeviciute - Fixed comment parent XSS vulnerability. The 20-character preview wasn't being sanitized. Added markdown filter:
   ```django
   {{ comment.parent.body_text|markdown|truncatechars:20 }}
   ```
   Also added test_comment_parent_xss_protection to verify this fix.

5. @karina-klinkeviciute - Added test_comment_xss_end_to_end_protection integration test. Posts a comment with XSS payload and verifies the rendered page is properly sanitized.

6. Added ARIA attribute support (aria-label, aria-describedby, aria-hidden, etc.) - required for government site accessibility compliance. Added test_comment_preserves_aria_attributes and test_sanitize_html_preserves_aria tests.

7. Added rel="nofollow noopener noreferrer" to links via set_safe_link_attrs callback. Added test_sanitize_html_adds_safe_link_attributes test.

8. @karina-klinkeviciute - Comments on projects/use cases: These use the same component/comment.html template, so they're already protected by this fix.

## Deferred to Separate Issues

9. Task descriptions, datasets, organizations: Similar XSS patterns exist in other modules. Will create separate issues to address it.

10. CSP headers: Requires infrastructure/middleware changes and testing across entire application. Will create separate issue with implementation plan.

LinasRaj committed:
Merge branch 'devel' into 254-fix-comments-xss

LinasRaj committed:
Updated poetry lock file

LinasRaj commented:
https://254-fix-comments-xss.branch-test.data.gov.lt/
